src/main/java/com/fasterxml/jackson/databind/ser/std/MapSerializer.java

```diff
@@ -1166,23 +1167,57 @@ protected Map<?,?> _orderEntries(Map<?,?> input, JsonGenerator gen,
         if (input instanceof SortedMap<?,?>) {
             return input;
         }
-        // [databind#1411]: TreeMap does not like null key... (although note that
-        //   check above should prevent this code from being called in that case)
-        // [databind#153]: but, apparently, some custom Maps do manage hit this
-        //   problem.
-        if (_hasNullKey(input)) {
-            TreeMap<Object,Object> result = new TreeMap<Object,Object>();
-            for (Map.Entry<?,?> entry : input.entrySet()) {
-                Object key = entry.getKey();
-                if (key == null) {
-                    _writeNullKeyedEntry(gen, provider, entry.getValue());
-                    continue;
+        // or is it empty? then no need to sort either
+        if (input == null || input.isEmpty()) {
+            return input;
+        }
+        // [databind#4773] Since 2.19: We should not try sorting Maps with uncomparable keys
+        // And first key is a good enough sample for now.
+        Object firstKey = input.keySet().iterator().next();
+        if (firstKey != null && !Comparable.class.isInstance(firstKey)) {
+            // We cannot sort incomparable keys, should we fail or just skip sorting?
+            if (!provider.isEnabled(SerializationFeature.FAIL_ON_ORDER_MAP_BY_INCOMPARABLE_KEY)) {
+                return input;
+            } else {
+                provider.reportBadDefinition(firstKey.getClass(),
+                    String.format("Cannot order Map entries by key of incomparable type %s, consider disabling " +
+                                "SerializationFeature.FAIL_ON_ORDER_MAP_BY_INCOMPARABLE_KEY to simply skip sorting",
+                                ClassUtil.classNameOf(firstKey)));
+            }
+        }
+        try {
+            // [databind#1411]: TreeMap does not like null key... (although note that
+            //   check above should prevent this code from being called in that case)
+            // [databind#153]: but, apparently, some custom Maps do manage hit this
+            //   problem.
+            if (_hasNullKey(input)) {
+                TreeMap<Object,Object> result = new TreeMap<Object,Object>();
+                for (Map.Entry<?,?> entry : input.entrySet()) {
+                    Object key = entry.getKey();
+                    if (key == null) {
+                        _writeNullKeyedEntry(gen, provider, entry.getValue());
+                        continue;
+                    }
+                    result.put(key, entry.getValue());
                 }
-                result.put(key, entry.getValue());
+                return result;
+            }
+            return new TreeMap<Object,Object>(input);
+        } catch (ClassCastException cce) {
```

Comment:
Unfortunately the try-catching only just got bigger :-( 
Without try-catching like we do here or finding some other way, this feature will only be un-deterministic.

If we look at below test [testSerializationSuccessWhenDisabledWithMixedTypes()](https://github.com/FasterXML/jackson-databind/pull/4778/files#r1827845527) case, you can see this feature is so easy to violate what the name FAIL_ON_ORDER_MAP_BY_INCOMPARABLE_KEY itself imposes.

I am also thinking without figuring out how to handle the case with incomparable key in the middle, we might be better off not push through this feature 🤔

Lemme know how you see @cowtowncoder !  🙏🏼

Reply:
My initial reaction: 

1. I don't want any try-catch block; and I am ok with mixed-type case failing: it would be unsupported case.
2. The idea is to improve handling, and to me uniform key types is what can be supported for Java.
3. I don't have problem for having edge cases that are no better, as long as presumed common case is improved -- which I think it is.

So I don't think we really must solve all cases; especially when trying to do that leads to awkward and brittle reliance on JDK-thrown exceptions, which are basically side-effects of type casts.

But let me have a look at PR see if that changes my mind.

Reply:
> But let me have a look at PR see if that changes my mind.

You don't need to. 
You are right on saying "what can be supported" or not. Mixed-type case would be a painful case to handle from maintenance point of view, and also actual usage wouldn't be much reasonable either.

Reply:
Plus, if we change our mind, we can always add catch block -- or, even if we choose to, scan the whole keySet() to check type -- as a follow-up improvement.